### PR TITLE
Save/Reload games during draft or research phases

### PR DESCRIPTION
**Pull request by Lynesth**
This properly saves the game each time all players have drafted 1 card and passed the others to their neighbour.
So in case of the server going awol it will go back to the point it was last save in the draft.

It also saves when the draft is done (or if playing without the draft variant) while everyone is deciding which cards to buy (aka research phase). If server goes down, it'll go back to the beginning of the research phase, for all players.

Again, this will add a few saves per game - around 50 rows for a 4p game with all draft variants - but it seems negligible.

Quick summary of what the code consists of:
- A new method, `game.save()`, which saves the game \o/
- A refactor of `playerIsFinishedWithDraftingPhase` because it was a mess to understand. Still not 100% happy with it but I think it's better.
- A removal of `isLastActiveRoundOfDraft` (linked to the refactor above) and instead use the numbers of cards left to decide if it was the last round.
- A small reordering of things in `getDraftCardsFrom`, nothing important.
- A bunch of `this.save()` added here and there to save the game at appropriate times.
- A better handling of what to do when the game is loaded from json, so that it can restart during draft or research phase (the whole point of this PR).

-----

This was obviously ~~troughly~~ ~~throughouly~~ ~~thoroug...~~ very well tested by yours truly.

### CODE REVIEW

**Comment by Lynesth** [comment]
Closes https://github.com/bafolts/terraforming-mars/issues/2098
Fixes https://github.com/bafolts/terraforming-mars/issues/2109

File: src/Game.ts [hunk 1]
```diff
@@ -319,22 +320,11 @@ export class Game implements ISerializable<SerializedGame> {
...
-          }
         } else {
-          this.setStartingProductions(player);
           this.playerHasPickedCorporationCard(player, new BeginnerCorporation());
         }
       }
 
-      // Save initial game state
-      Database.getInstance().saveGameState(this.id, this.lastSaveId, this.toJSON(), this.players.length);
-      this.lastSaveId = 1;
```
Comment (Lynesth) [inline src/Game.ts:336]: Game is now saved when entering either `runDraftRound` or `gotoInitialResearchPhase` a few lines below this.

**Review by bafolts — COMMENTED** [review]
A quick read through this looks right. As this is dealing with some of the lower level stuff again I am going to take time to ensure we aren't missing something. Will look again tomorrow.

**Review by bafolts — APPROVED** [review]

**Comment by alrusdi** [comment]
@Lynesth feel free to merge it to main!

**Comment by lomky** [comment]
Much appreciated! :rocket:

**Pull request merged**